packages/headless/src/app/ssr-engine/types/common.ts

```diff
@@ -0,0 +1,110 @@
+import {AnyAction} from '@reduxjs/toolkit';
+import {Controller} from '../../../controllers';
+import {CoreEngine} from '../../engine';
+
+export type HasKeys<TObject> = TObject extends {}
+  ? keyof TObject extends never
+    ? false
+    : true
+  : boolean;
+
+export interface OptionsExtender<TOptions> {
+  (options: TOptions): TOptions | Promise<TOptions>;
+}
+
+export interface ControllersPropsMap {
+  [customName: string]: unknown;
+}
+
+export interface ControllersMap {
+  [customName: string]: Controller;
+}
+
+export interface ControllerSnapshot<TState> {
+  state: TState;
+}
```

Comment:
One thing I updated in the RFC after all was to rename "Snapshot" to "InitialState". I think it would be more intuitive since it's returned by `fetchInitialState`. Should probably update it in this PR too

Reply:
Ok, done.